pageserver/src/page_service.rs

```diff
@@ -137,6 +159,88 @@ pub fn spawn(
     Listener { cancel, task }
 }
 
+/// Spawns a gRPC server for the page service.
+pub fn spawn_grpc(
+    conf: &'static PageServerConf,
+    tenant_manager: Arc<TenantManager>,
+    auth: Option<Arc<SwappableJwtAuth>>,
+    perf_trace_dispatch: Option<Dispatch>,
+    listener: std::net::TcpListener,
+    tls_identity: Option<tonic::transport::Identity>,
+) -> anyhow::Result<CancellableTask> {
+    let cancel = CancellationToken::new();
+    let ctx = RequestContextBuilder::new(TaskKind::PageRequestHandler)
+        .download_behavior(DownloadBehavior::Download)
+        .perf_span_dispatch(perf_trace_dispatch)
+        .detached_child();
+    let gate = Gate::default();
+
+    // Set up the TCP socket. We take a preconfigured TcpListener to bind the
+    // port early during startup.
+    let incoming = {
+        let _runtime = COMPUTE_REQUEST_RUNTIME.enter(); // required by TcpListener::from_std
+        listener.set_nonblocking(true)?;
+        tonic::transport::server::TcpIncoming::from_listener(
+            tokio::net::TcpListener::from_std(listener)?,
+            GRPC_TCP_NODELAY,
+            Some(GRPC_TCP_KEEPALIVE_TIME),
+        )
+        .map_err(|e| anyhow!("{e}"))
+    }?;
+
+    // Set up the gRPC server.
+    //
+    // TODO: consider tuning window sizes.
+    // TODO: wire up tracing.
+    let mut server = tonic::transport::Server::builder()
+        .http2_keepalive_interval(Some(GRPC_HTTP2_KEEPALIVE_INTERVAL))
+        .http2_keepalive_timeout(Some(GRPC_HTTP2_KEEPALIVE_TIMEOUT))
+        .max_concurrent_streams(Some(GRPC_MAX_CONCURRENT_STREAMS));
+    if let Some(identity) = tls_identity {
+        // TODO: unlike the HTTPS server, the gRPC server does not support
+        // certificate reloading. If this is needed, we have to use e.g.
+        // tonic-rustls, but it's unclear if it's mature enough.
+        server = server.tls_config(tonic::transport::ServerTlsConfig::new().identity(identity))?;
```

Comment:
@DimasKovas What's the background on `ReloadingCertificateResolver`? Tonic doesn't natively expose the rustls API, so we'll basically have to reimplement the server loop, like in this example:

https://github.com/hyperium/tonic/blob/master/examples/src/tls_rustls/server.rs

I take it this will be a hard requirement?

Reply:
We already have this loop for grpc protocol in storage broker:
https://github.com/neondatabase/neon/blob/main/storage_broker/src/bin/storage_broker.rs#L740

But in storage broker we use a separate implementation because we need to handle grpc and HTTP metric requests at the same port.

If you only need grpc on this port, I guess you can use https://github.com/neondatabase/neon/blob/main/libs/http-utils/src/server.rs#L25

Unfortunately, tonic's ServerTlsConfig sucks because it does not support live certificate reload. So we can't use tonic's Server

Reply:
Looks like the `http-utils` server is built on Hyper 0. We're going to be leaning heavily on HTTP/2 here, in the read hot path, so I'd like to stay reasonably current from the start.

Sounds like I'll basically have to reimplement the Tonic server using Hyper1? This feels pretty heavy-handed. Is this really worth the maintenance cost just for the TLS reloading?

Reply:
I'm going to punt on TLS for now. I've added an item to https://github.com/neondatabase/neon/issues/11728.